examples/go/vm-reconfig-via-tag/handler/handler.go

```diff
@@ -0,0 +1,564 @@
+package function
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+
+	handler "github.com/openfaas/templates-sdk/go-http"
+	"github.com/pelletier/go-toml"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vapi/rest"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+const cfgPath = "/var/openfaas/secrets/vcconfig"
+
+var (
+	vsClt *vsClient // Persist vSphere connection.
+	once  sync.Once // For handleSignal() to be called once.
+	debug bool      // True for verbose logging.
+)
+
+// vcConfig represents the toml vcconfig file
+type vcConfig struct {
+	VCenter struct {
+		Server, User, Password string
+		Insecure               bool
+	}
+}
+
+// cloudEvent is a subsection of a Cloud Event.
+type cloudEvent struct {
+	Data types.Event `json:"data,omitempty"`
+}
+
+type vsClient struct {
+	govmomi *govmomi.Client
+	rest    *rest.Client
+	tagMgr  *tags.Manager
+}
+
+type vmConfig struct {
+	name    string
+	valBool bool
+	valInt  int
+}
+
+// Handle a function invocation
+func Handle(req handler.Request) (handler.Response, error) {
+	ctx := context.Background()
+	debug = setVerbosity()
+
+	// Load config every time, to ensure the most updated version is used.
+	cfg, err := loadConfig(cfgPath)
+	if err != nil {
+		return handlerResponseWithError("load vcconfig: %w",
+			http.StatusInternalServerError,
+			err,
+		)
+	}
+
+	// The Mananged Object Reference for the VM that powered on.
+	vmMOR, err := eventVmMOR(req.Body)
+	if err != nil {
+		return handlerResponseWithError(
+			"retrieve VM object: %w",
+			http.StatusInternalServerError,
+			err,
+		)
+	}
+
+	if vsClt == nil {
+		// Save vSphere connection in a global variable.
+		vsClt, err = newClient(ctx, cfg)
+		if err != nil {
+			return handlerResponseWithError(
+				"connect to vSphere: %w",
+				http.StatusInternalServerError,
+				err,
+			)
+		}
+	}
+
+	// Signals wil cause vSphere APIs to log out.
+	once.Do(func() {
+		go handleSignal(ctx, vsClt)
+	})
+
+	// Check REST connection by getting tag objects.
+	tagObjs, err := vsClt.attachedTags(ctx, *vmMOR)
+	if err != nil {
+		// Retry once if not connected after creating new client.
+		if strings.Contains(err.Error(), "401 Unauthorized") {
```

Comment:
Assuming there's not typed error we can use here, right?

Reply:
That correct. In Govmomi's tags_association.go line 88, an http.MethodPost call is made to retrieve the list of attached tags. The error is returned from the API endpoint and doesn't seem to have an associated typed error (I could be mistaken).

```
func (c *Manager) ListAttachedTags(ctx context.Context, ref mo.Reference) ([]string, error) {
	spec := internal.NewAssociation(ref)
	url := c.Resource(internal.AssociationPath).WithAction("list-attached-tags")
	var res []string
	return res, c.Do(ctx, url.Request(http.MethodPost, spec), &res)
}
```